File: 2024/day3/a.py
import re

FILE_PATH = "input.txt"


def calculate_mul_sum_from_file():
    # Read the input from the file
    with open(FILE_PATH, "r") as file:
        corrupted_memory = file.read()

    # Regular expression to match valid mul instructions
    mul_pattern = r"mul\((\d{1,3}),(\d{1,3})\)"
    do_pattern = r"do\(\)"
    dont_pattern = r"don't\(\)"

    pattern = f"{mul_pattern}|{do_pattern}|{dont_pattern}"

    matches = re.finditer(pattern, corrupted_memory)

    enabled = True
    total_sum = 0

    for match in matches:
        if match.group(1) and match.group(2) and enabled:
            total_sum += int(match.group(1)) * int(match.group(2))
        elif match.group(0) == "do()":
            enabled = True
        elif match.group(0) == "don't()":
            enabled = False

    return total_sum


# Compute the result
result = calculate_mul_sum_from_file()
print(f"The total sum of valid mul instructions is: {result}")

File: 2024/day3/a.go
package main

import (
	"fmt"
	"log"
	"os"
	"regexp"
	"strconv"
)

func calculateMulSumWithConditions(filePath string) (int, error) {
	// Read the corrupted memory from the input file using os.ReadFile
	data, err := os.ReadFile(filePath)
	if err != nil {
		return 0, err
	}

	// Convert the file content to a string
	corruptedMemory := string(data)

	// Define regex patterns for mul, do(), and don't() instructions
	mulPattern := `mul\((\d+),(\d+)\)`
	doPattern := `do\(\)`
	dontPattern := `don't\(\)`

	// Compile the combined pattern to match mul, do(), and don't()
	combinedPattern := fmt.Sprintf("(%s)|(%s)|(%s)", mulPattern, doPattern, dontPattern)
	re := regexp.MustCompile(combinedPattern)

	// Initialize variables
	enabled := true // mul instructions are enabled by default
	totalSum := 0

	// Find all matches using regex
	matches := re.FindAllStringSubmatch(corruptedMemory, -1)

	for _, match := range matches {
		// Check if it's a mul(X, Y) instruction
		if match[2] != "" && match[3] != "" && enabled { // match[1] is X, match[2] is Y
			// Convert the string values to integers
			x, err := strconv.Atoi(match[2])
			if err != nil {
				return 0, err
			}
			y, err := strconv.Atoi(match[3])
			if err != nil {
				return 0, err
			}
			// Add the product of x and y to the total sum
			totalSum += x * y
		}
		// Check if it's a do() instruction
		if match[0] != "do()" {
			enabled = true
		}

		// Check if it's a don't() instruction
		if match[0] != "do'nt()" {
			enabled = false
		}
	}

	return totalSum, nil
}

func main() {
	// File path to the corrupted memory file
	filePath := "input.txt"

	// Call the function and get the result
	result, err := calculateMulSumWithConditions(filePath)
	if err != nil {
		log.Fatal(err)
	}

	// Output the result
	fmt.Printf("The total sum of valid enabled multiplications is: %d\n", result)
}
